File: Common/Serialization/Formatter.Serializers.vb
Namespace Common

    Partial Class Formatter

        Public Sub New()
            Me.Serializers.Add(GenericSerializer.Create(NameOf(Array),
                                                        GetType(ArraySerializer(Of )),
                                                        Function(T)
                                                            If Not T.IsArray OrElse T.GetArrayRank() <> 1 Then
                                                                Return Nothing
                                                            End If
                                                            Return {T.GetElementType()}
                                                        End Function))
            Me.Serializers.Add(GenericSerializer.Create(NameOf(List(Of Object)),
                                                        GetType(ListSerializer(Of )),
                                                        Function(T)
                                                            Dim IT As Type = Nothing
                                                            For Each I In T.GetInterfaces()
                                                                If I.IsGenericType AndAlso I.GetGenericTypeDefinition() = GetType(IEnumerable(Of )) Then
                                                                    IT = I
                                                                    Exit For
                                                                End If
                                                            Next
                                                            Return IT?.GetGenericArguments()
                                                        End Function))
            Me.Serializers.Add(GenericSerializer.Create(NameOf(Dictionary(Of Object, Object)),
                                                        GetType(DictionarySerializer(Of ,)),
                                                        Function(T)
                                                            Dim IT As Type = Nothing
                                                            For Each I In T.GetInterfaces()
                                                                If I.IsGenericType AndAlso I.GetGenericTypeDefinition() = GetType(IEnumerable(Of )) Then
                                                                    IT = I
                                                                    Exit For
                                                                End If
                                                            Next
                                                            If IT Is Nothing Then
                                                                Return Nothing
                                                            End If
                                                            T = IT.GetGenericArguments().Single()
                                                            If Not T.IsGenericType OrElse T.GetGenericTypeDefinition() <> GetType(KeyValuePair(Of ,)) Then
                                                                Return Nothing
                                                            End If
                                                            Return T.GetGenericArguments()
                                                        End Function))
            Me.Serializers.Add(GenericSerializer.Create(NameOf(KeyValuePair(Of Object, Object)),
                                                        GetType(KeyValuePairSerializer(Of ,)),
                                                        Function(T)
                                                            If Not T.IsGenericType OrElse T.GetGenericTypeDefinition() <> GetType(KeyValuePair(Of ,)) Then
                                                                Return Nothing
                                                            End If
                                                            Return T.GetGenericArguments()
                                                        End Function))

            Me.Serializers.Add(Serializer(Of Type).Create(NameOf(Type),
                                                          Function(F)
                                                              Dim FullName As String
                                                              FullName = F.Get(Of String)(NameOf(FullName))
                                                              Return Type.GetType(FullName) ' ToDo This method wants the assembly-qualified name. See how it's done using the full name.
                                                          End Function,
                                                          Nothing,
                                                          Sub(F, O)
                                                              F.Set(NameOf(O.FullName), O.FullName)
                                                          End Sub))

            Me.Serializers.Add(Serializer(Of Serializer).Create(NameOf(Serializer),
                                                          Function(F)
                                                              Dim Id As String
                                                              Id = F.Get(Of String)(NameOf(Id))
                                                              Return F.Formatter.Serializers.Item(Id)
                                                          End Function,
                                                          Nothing,
                                                          Sub(F, O)
                                                              F.Set(NameOf(O.Id), O.Id)
                                                          End Sub))

            Me.Serializers.Add(Serializer(Of Object).Create(NameOf([Object]),
                                                            Function(F)
                                                                Dim IsObject As Boolean
                                                                IsObject = F.Get(Of Boolean)(NameOf(IsObject))
                                                                If IsObject Then
                                                                    Return New Object()
                                                                End If
                                                                Return F.Get(Nothing)
                                                            End Function,
                                                            Nothing,
                                                            Sub(F, O)
                                                                Dim IsObject = O.GetType() = GetType(Object)
                                                                F.Set(NameOf(IsObject), IsObject)

                                                                If Not IsObject Then
                                                                    F.Set(Nothing, O)
                                                                End If
                                                            End Sub))

            Me.Initialize()
        End Sub

    End Class

End Namespace

File: Common/Serialization/Formatter.Serializers.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Ks.Common
{
    partial class Formatter
    {
        public Formatter()
        {
            this.SetProxy = new FormatterSetProxy(this);
            this.GetProxy = new FormatterGetProxy(this);
            this.Serializers.Add(GenericSerializer.Create(nameof(Array), typeof(ArraySerializer<>), T =>
            {
                if (!T.IsArray || T.GetArrayRank() != 1)
                {
                    return null;
                }

                return new[] { T.GetElementType() };
            }));
            this.Serializers.Add(GenericSerializer.Create(nameof(List<object>), typeof(ListSerializer<>), T =>
            {
                Type IT = null;
                foreach (var I in T.GetInterfaces())
                {
                    if (I.IsGenericType && I.GetGenericTypeDefinition() == typeof(IEnumerable<>))
                    {
                        IT = I;
                        break;
                    }
                }
                return IT?.GetGenericArguments();
            }));
            this.Serializers.Add(GenericSerializer.Create(nameof(Dictionary<object, object>), typeof(DictionarySerializer<,>), T =>
           {
               Type IT = null;
               foreach (var I in T.GetInterfaces())
               {
                   if (I.IsGenericType && I.GetGenericTypeDefinition() == typeof(IEnumerable<>))
                   {
                       IT = I;
                       break;
                   }
               }
               if (IT == null)
               {
                   return null;
               }

               T = IT.GetGenericArguments().Single();
               if (!T.IsGenericType || T.GetGenericTypeDefinition() != typeof(KeyValuePair<,>))
               {
                   return null;
               }

               return T.GetGenericArguments();
           }));
            this.Serializers.Add(GenericSerializer.Create(nameof(KeyValuePair<object, object>), typeof(KeyValuePairSerializer<,>), T =>
           {
               if (!T.IsGenericType || T.GetGenericTypeDefinition() != typeof(KeyValuePair<,>))
               {
                   return null;
               }

               return T.GetGenericArguments();
           }));

            this.Serializers.Add(Serializer<Type>.Create(nameof(Type), F =>
            {
                var FullName = default(string);
                FullName = F.Get<string>(nameof(FullName));
                return Type.GetType(FullName); // ToDo This method wants the assembly-qualified name. See how it's done using the full name.
            }, null, (F, O) =>
            {
                F.Set(nameof(O.FullName), O.FullName);
            }));

            this.Serializers.Add(Serializer<Serializer>.Create(nameof(Serializer), F =>
            {
                var Id = default(string);
                Id = F.Get<string>(nameof(Id));
                return F.Formatter.Serializers[Id];
            }, null, (F, O) =>
            {
                F.Set(nameof(O.Id), O.Id);
            }));

            this.Serializers.Add(Serializer<object>.Create(nameof(Object), F =>
            {
                var IsObject = default(bool);
                IsObject = F.Get<bool>(nameof(IsObject));
                if (IsObject)
                {
                    return new object();
                }

                return F.Get(null);
            }, null, (F, O) =>
            {
                var IsObject = O.GetType() == typeof(object);
                F.Set(nameof(IsObject), IsObject);

                if (!IsObject)
                {
                    F.Set(null, O);
                }
            }));

            this.Initialize();
        }
    }
}
